### ISSUE TITLE
5.43.0   translate / compile error

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#3339]
Imported from SourceForge on 2024-07-07 20:56:22
Created by **[thunk06](https://sourceforge.net/u/thunk06/)** on 2019-06-02 09:13:46
Original: https://sourceforge.net/p/maxima/bugs/3556

---

I have just dowloaded the brand new maxima-5.43.0-win64   ( sbcl )

As usual, with each new release, I am recompiling all my stuff to generate "fasl"

So, I do have with this release, several errors reported that I never had in the past
For example with the simple simple function in attachement that I have extracted 
from one of my mac file

The UNLISP file reports 

note: translating addforks
warning: encountered undefined variable info in translation.
warning: encountered undefined variable node in translation.
warning: encountered undefined variable choices in translation.
warning: encountered undefined variable A2B in translation.
warning: encountered undefined variable B2A in translation.
translator: internal error. Message: "Bad lisp expression generated."
warning: encountered undefined variable list_forks in translation.
error: failed to translate addforks
note: keep going and hope for the best.

I do not know why list_forks is now in trouble since it is defined
and initialized at the first line of the file 

Regards,
    Jacques

Attachments:
- https://sourceforge.net/p/maxima/bugs/3556/attachment/addforks.mac

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 20:56:24
Created by **[thunk06](https://sourceforge.net/u/thunk06/)** on 2019-06-02 16:20:33
Original: https://sourceforge.net/p/maxima/bugs/3556/#2a32

---

I am continuing to dig the problem

It seems that the translation to lisp does not like anymore the statement

                     list_forks : map( lambda( [ forks ] , .... ) , list_forks )
               
However, if  I define indepently a fonction and then use it within the map the 
translator is able to complete the addforks function without issuing an internal error

Unfortunatly I do use  map( lambda( ... ) , ...  in various files of my stuff

I do not know yet if this problem is the only one breaking now my (5K lines) analog 
electronic package 

Regards,
     Jacques
    
PS1:   my target is not  lisp  but  fasl  to benefit the speed of sbcl 
PS2:   with several Maxima past releases, I did not have any trouble

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 20:56:27
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2019-06-08 19:38:41
Original: https://sourceforge.net/p/maxima/bugs/3556/#4cf7

---

From what I can tell, the problem is that the Maxima-to-Lisp translator is incorrectly handling a construct which was introduced recently, to set the global flag `prederror` to `true` in translated code. (`prederror = true` causes an error if an `if` condition evaluates to something other than `true` or `false`; this is the best option for translated code.)

I am working on an improvement for the translator to handle the construct which contains `prederror`. (It is a Common Lisp lexical block, (LET (($PREDERROR T)) ...) to be precise. The problem is not `lambda` or even `prederror` but rather the handling of LET in generated code.)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 20:56:31
Created by **[thunk06](https://sourceforge.net/u/thunk06/)** on 2019-06-09 17:19:38
Original: https://sourceforge.net/p/maxima/bugs/3556/#97ca

---

Thanks a lot for looking at this ticket
I am a long time Maxima user and I am following you regularly on maxima-discuss

I am in a middle of something and I cannot these days simplify your debug by doing a divide
and conquer slicing of my 5K lines code

However, here is in attachement the 4 fonctions that do not translate since 5.43.0

Best Regards,
      Jacques



Attachments:
- https://sourceforge.net/p/maxima/bugs/_discuss/thread/d54c7edbd6/97ca/attachment/translate_errors.mac

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 20:56:34
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2019-06-09 20:51:07
Original: https://sourceforge.net/p/maxima/bugs/3556/#653e

---

I've fixed this bug with commit 95ded71, therefore closing this report as fixed.

I've copied that commit to the Maxima 5.43 release branch, and if there is another 5.43 release (i.e. 5.43.1), it will be included. Otherwise, this bug fix will be present in the 5.44 release, which will presumably be released some months from now.

The essential part of the commit is the following Lisp code. One could copy this code to a file, let's say `free-lisp-vars.lisp`, and load it via `load("free-lisp-vars.lisp");`. Then the translation of "if" within lambda should work again.
~~~~
(defun-prop (let free-lisp-vars) (form)
  (union-var-set 
    (list
      ;; extract (FOO BAR NIL NIL) from (LET ((A FOO) (B BAR) C D) ...)
      ;; and apply FREE-LISP-VARS to each.
      (union-var-set (mapcar #'free-lisp-vars (mapcar #'(lambda (e) (if (consp e) (cdr e))) (cadr form))))
      (difference-var-sets
        ;; cargo-cult programming: copy this next bit from (DEFUN-PROP (PROG ...)) above.
        (union-var-set
          (mapcar #'(lambda (u)
                      (cond ((atom u) nil) ;; go tag.
                            (t
                              (free-lisp-vars u))))
                  (cddr form)))
        ;; extract A B C D from (LET ((A FOO) (B BAR) C D) ...)
        (make-var-set (mapcar #'(lambda (e) (if (atom e) e (car e))) (cadr form)))))))
~~~~

Jacques, I applied `translate_file` to your `translate_errors.mac`, and `translate_file` did not report any errors (now that I have made the bug fix). If you have test cases so that you can verify the result of translated code, that would be very interesting to try.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-07 20:56:38
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2019-06-09 20:52:08
Original: https://sourceforge.net/p/maxima/bugs/3556/#5910

---

- **labels**:  --> translator, translate, translate_file, prederror
- **status**: open --> closed
